### PR TITLE
Pass WebSocketBuilder to HttpApp in builders

### LINKED ISSUE #5108: [ws] Any effect you like, as long as it's IO
We seem to make an assumption that we're running in `IO`, based on the hardcoded key [here](https://github.com/http4s/http4s/blob/a942a83e2b88bafc644fc9916fd1979eb951ced8/server/shared/src/main/scala/org/http4s/server/websocket/package.scala#L26-L30).

This can be exposed if your routes are defined in an effect that differs from the effect the server is running in, as via the `liftR` operation provided by [natchez-http4s](https://github.com/tpolecat/natchez-http4s), which turns an `HttpRoutes[Kleisli[F, Span[F], *]]` into an `HttpRoutes[F]` via a pair of natural transformations. A simplified version here exhibits the same problem:

```scala
package test

import cats._
import cats.data.Kleisli
import cats.effect._
import org.http4s.HttpRoutes
import org.http4s.blaze.server.BlazeServerBuilder
import org.http4s.dsl.Http4sDsl
import org.http4s.implicits._
import org.http4s.server.websocket.WebSocketBuilder
import org.http4s.websocket.WebSocketFrame.Text
import scala.concurrent.ExecutionContext

object Test extends IOApp.Simple {

  type K[A] = Kleisli[IO,Int,A]

  // Our routes run in Kleisli
  val routes: HttpRoutes[K] = {
    object dsl extends Http4sDsl[K]; import dsl._
    HttpRoutes.of[K] {
      case GET -> Root / "ws" =>
        WebSocketBuilder[K].build(_.map(f => Text(s"You said $f")))
    }
  }

  // We can unlift our routes back to IO
  def unlift(routes: HttpRoutes[K]): HttpRoutes[IO] =
    Kleisli { ioReq =>
      val io2k: IO ~> K = Kleisli.liftK
      val k2io: K ~> IO = Kleisli.applyK(42)
      routes.run(ioReq.mapK(io2k)).mapK(k2io).map(_.mapK(k2io))
    }

  // Main app and server are in IO
  def run: IO[Unit] =
    BlazeServerBuilder[IO](ExecutionContext.global)
      .withHttpApp(unlift(routes).orNotFound)
      .bindHttp(8080, "0.0.0.0")
      .resource
      .use(_ => IO.never)

}
```

Using `websocat` to connect and send a message yields

```
$ websocat ws://localhost:8080/ws
hi

websocat: WebSocketError: I/O failure
websocat: error running
$
```

And over on the console we see

```
[INFO] org.http4s.blaze.channel.nio1.NIO1SocketServerGroup - Service bound to address /0:0:0:0:0:0:0:0:8080
[INFO] org.http4s.blaze.server.BlazeServerBuilder -
  _   _   _        _ _
 | |_| |_| |_ _ __| | | ___
 | ' \  _|  _| '_ \_  _(_-<
 |_||_\__|\__| .__/ |_|/__/
             |_|
[INFO] org.http4s.blaze.server.BlazeServerBuilder - http4s v0.15.1 on blaze v0.15.1 started at http://[::]:8080/
[INFO] org.http4s.blaze.channel.nio1.NIO1SocketServerGroup - Accepted connection from /0:0:0:0:0:0:0:1:55117
[ERROR] org.http4s.blaze.server.WSFrameAggregator - Exception caught when attempting inbound command
java.lang.ClassCastException: class cats.data.Kleisli cannot be cast to class cats.effect.IO (cats.data.Kleisli and cats.effect.IO are in unnamed module of loader 'app')
	at cats.effect.IO$$anon$2.attempt(IO.scala:1268)
	at cats.syntax.ApplicativeErrorOps$.attempt$extension(applicativeError.scala:52)
	at org.http4s.blazecore.websocket.Http4sWSStage.stageStartup(Http4sWSStage.scala:174)
	at org.http4s.blaze.pipeline.Stage.inboundCommand(Stages.scala:85)
	at org.http4s.blaze.pipeline.Stage.inboundCommand$(Stages.scala:83)
	at org.http4s.blazecore.websocket.Http4sWSStage.inboundCommand(Http4sWSStage.scala:44)
	at org.http4s.blaze.pipeline.Head.sendInboundCommand(Stages.scala:268)
	at org.http4s.blaze.pipeline.Head.sendInboundCommand$(Stages.scala:264)
	at org.http4s.blaze.server.WSFrameAggregator.sendInboundCommand(WSFrameAggregator.scala:32)
	at org.http4s.blaze.pipeline.Head.inboundCommand(Stages.scala:282)
	at org.http4s.blaze.pipeline.Head.inboundCommand$(Stages.scala:280)
	at org.http4s.blaze.server.WSFrameAggregator.inboundCommand(WSFrameAggregator.scala:32)
	at org.http4s.blaze.pipeline.Head.sendInboundCommand(Stages.scala:268)
	at org.http4s.blaze.pipeline.Head.sendInboundCommand$(Stages.scala:264)
	at org.http4s.blaze.server.WebSocketDecoder.sendInboundCommand(WebSocketDecoder.scala:25)
	at org.http4s.blaze.pipeline.Head.inboundCommand(Stages.scala:282)
	at org.http4s.blaze.pipeline.Head.inboundCommand$(Stages.scala:280)
	at org.http4s.blaze.server.WebSocketDecoder.inboundCommand(WebSocketDecoder.scala:25)
	at org.http4s.blaze.pipeline.Head.sendInboundCommand(Stages.scala:268)
	at org.http4s.blaze.pipeline.Head.sendInboundCommand$(Stages.scala:264)
	at org.http4s.blazecore.IdleTimeoutStage.sendInboundCommand(IdleTimeoutStage.scala:29)
	at org.http4s.blaze.pipeline.Tail.replaceTail(Stages.scala:193)
	at org.http4s.blaze.pipeline.Tail.replaceTail$(Stages.scala:172)
	at org.http4s.blaze.server.Http1ServerStage.replaceTail(Http1ServerStage.scala:86)
	at org.http4s.blaze.server.WebSocketSupport.$anonfun$renderResponse$7(WebSocketSupport.scala:109)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:448)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
[ERROR] org.http4s.blaze.channel.nio1.NIO1HeadStage - Abnormal NIO1HeadStage termination
java.lang.ClassCastException: class cats.data.Kleisli cannot be cast to class cats.effect.IO (cats.data.Kleisli and cats.effect.IO are in unnamed module of loader 'app')
	at cats.effect.IO$$anon$2.attempt(IO.scala:1268)
	at cats.syntax.ApplicativeErrorOps$.attempt$extension(applicativeError.scala:52)
	at org.http4s.blazecore.websocket.Http4sWSStage.stageStartup(Http4sWSStage.scala:174)
	at org.http4s.blaze.pipeline.Stage.inboundCommand(Stages.scala:85)
	at org.http4s.blaze.pipeline.Stage.inboundCommand$(Stages.scala:83)
	at org.http4s.blazecore.websocket.Http4sWSStage.inboundCommand(Http4sWSStage.scala:44)
	at org.http4s.blaze.pipeline.Head.sendInboundCommand(Stages.scala:268)
	at org.http4s.blaze.pipeline.Head.sendInboundCommand$(Stages.scala:264)
	at org.http4s.blaze.server.WSFrameAggregator.sendInboundCommand(WSFrameAggregator.scala:32)
	at org.http4s.blaze.pipeline.Head.inboundCommand(Stages.scala:282)
	at org.http4s.blaze.pipeline.Head.inboundCommand$(Stages.scala:280)
	at org.http4s.blaze.server.WSFrameAggregator.inboundCommand(WSFrameAggregator.scala:32)
	at org.http4s.blaze.pipeline.Head.sendInboundCommand(Stages.scala:268)
	at org.http4s.blaze.pipeline.Head.sendInboundCommand$(Stages.scala:264)
	at org.http4s.blaze.server.WebSocketDecoder.sendInboundCommand(WebSocketDecoder.scala:25)
	at org.http4s.blaze.pipeline.Head.inboundCommand(Stages.scala:282)
	at org.http4s.blaze.pipeline.Head.inboundCommand$(Stages.scala:280)
	at org.http4s.blaze.server.WebSocketDecoder.inboundCommand(WebSocketDecoder.scala:25)
	at org.http4s.blaze.pipeline.Head.sendInboundCommand(Stages.scala:268)
	at org.http4s.blaze.pipeline.Head.sendInboundCommand$(Stages.scala:264)
	at org.http4s.blazecore.IdleTimeoutStage.sendInboundCommand(IdleTimeoutStage.scala:29)
	at org.http4s.blaze.pipeline.Tail.replaceTail(Stages.scala:193)
	at org.http4s.blaze.pipeline.Tail.replaceTail$(Stages.scala:172)
	at org.http4s.blaze.server.Http1ServerStage.replaceTail(Http1ServerStage.scala:86)
	at org.http4s.blaze.server.WebSocketSupport.$anonfun$renderResponse$7(WebSocketSupport.scala:109)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:448)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

Thanks @ChristopherDavenport for helping me track this down.

### PR DESCRIPTION
**Pull request by tpolecat**
This resolves #5108 by changing the builders to provide a compatible `WebSocketBuilder[F]` that will use the same vault `Key` as the server. The API changes are:

- The only way to get a `WebSocketBuilder` is from your server builder, by calling `.withHttpWebSocketApp(f: WebSocketBuilder[F] => HttpApp[F])`.
- `WebSocketBuilder` now has `imapK`, which allows downstream users to adapt it to contexts in which the local effect differs.

Another change is that Blaze has an annoying `webSocketEnabled` boolean, which (if it's false) will cause your app to blow up if you try to use websockets. This boolean saves a grand total of one hash lookup per request so I removed it as a simplification.

Waiting on:
- [x] https://github.com/typelevel/vault/pull/267

### CODE REVIEW

**Comment by rossabaker** [comment]
I took care of releasing and upgrading to a proper Vault, but there are MiMa failures galore.

**Comment by rossabaker** [comment]
So, yeah, adding the parameter to the case class is going to be nasty from a binary compatibility perspective.  We might need to rename `WebSocketBuilder` and deprecate.

**Comment by rossabaker** [comment]
I got this most of the way, but the new `WebSocketBuilder` value is blocking this on 0.23.

**Comment by tpolecat** [comment]
Ah, yeah I didn't even try to do this in a non-breaking way. I'll see what I can do.

**Comment by tpolecat** [comment]
say hello to my little `WebSocketBuilder2` :-\

**Comment by tpolecat** [comment]
Looks like everything passes, modulo flaky Blaze test.

File: server/src/main/scala/org/http4s/server/websocket/WebSocketBuilder2.scala [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+/** Build a response which will accept an HTTP websocket upgrade request and initiate a websocket connection using the
+  * supplied exchange to process and respond to websocket messages.
+  * @param headers Handshake response headers, such as such as:Sec-WebSocket-Protocol.
+  * @param onNonWebSocketRequest The status code to return to a client making a non-websocket HTTP request to this route.
+  *                              default: NotImplemented
+  * @param onHandshakeFailure The status code to return when failing to handle a websocket HTTP request to this route.
+  *                           default: BadRequest
+  */
+final case class WebSocketBuilder2[F[_]: Applicative](
```
Comment (rossabaker) [inline server/src/main/scala/org/http4s/server/websocket/WebSocketBuilder2.scala:40]: Making this a case class is how we got pinned in the first place.  Case class builders are bad, and not used most other places.  I'll PR a counterproposal.

**Comment by tpolecat** [comment]
scalafmt says no

**Review by rossabaker — APPROVED** [review]
I don't love the `WebSocketBuilder2` name, but I don't see a way out.  I think this is ready.

**Comment by adamw** [comment]
FYI, I think that while the change is binary-compatible in the end as far as compilation/classloading goes, it seems it did break the "old" way of using websockets (see the linked tapir issue - https://github.com/softwaremill/tapir/issues/1566). So maybe it's safe to create the source-level incompatibility, at least things would fail at compile-time :)

**Comment by rossabaker** [comment]
Thanks.  I opened a new issue to remind us to test it directly.  The deprecated path should still work, as long as there aren't any translations like prompted the bug in the first place.

**Pull request merged**